Tutorials/WinForms/Classic/implement-crud-functionality-with-xpo-objects.md

```diff
@@ -0,0 +1,163 @@
+* Close the both windows and open the `EditCustomerForm` in the designer. Drag a `SimpleButton` component from the toolbox and drop it to the `LayoutControl`. Add `EmptySpaceItem` elements to neatly align the button.
+* Rename the `simpleButton1` control to **btnSave** and change its `Text` to **&Save** (the ampersand sign assigns a mnemonic command to the button).
+* Double click the `btnSave` control to add the `Click` event handler (you can use the **Properties** window to add it).
+* Add this code to the `btnSave_Click` event handler:
+    ```csharp
+    private void btnSave_Click(object sender, EventArgs e) {
+        UnitOfWork.CommitChanges();
+        Close();
+    }
+    ```
+* The [UnitOfwork.CommitChanges](https://docs.devexpress.com/XPO/DevExpress.Xpo.UnitOfWork.CommitChanges) method saves changes to the database. To show changes in the Data Grid (`CustomersListForm`), change code as follows: 
+    ```csharp
+    private void Form1_Load(object sender, EventArgs e) {
+        Reload();
+    }
+    private void CustomersGridView_RowClick(object sender, RowClickEventArgs e) {
+        if (e.Clicks == 2) {
+            e.Handled = true;
+            int customerID = (int)CustomersGridView.GetRowCellValue(e.RowHandle, colOid);
+            using (EditCustomerForm form = new EditCustomerForm(customerID)) {
+                form.ShowDialog(this);
+                Reload();
+                CustomersGridView.FocusedRowHandle = CustomersGridView.LocateByValue("Oid", form.CustomerID.Value);
+            }
+        }
+    }
+    private void Reload() {
+        CustomersBindingSource.DataSource = new XPCollection<Customer>(new Session());
+    }
+    ```
+* Optionally, use the Visual Studio **Refactor** tool to rename the `Form1_Load` event handler to `CustomersListForm_Load`. Put the cursor at the method name and click the **Edit > Refactor > Rename** menu item or use **Ctrl+R,Ctrl+R**.   
+* Run the application, open the edit Form, change something, and click the **Save** button to see the result.
+## Handle concurrent changes    
+### Test with multiple users
+* Open the *App.config* file and change the connection string to use your database server ([K18445 - How to create a correct connection string for XPO providers](https://www.devexpress.com/Support/Center/Question/Details/K18445)). *You can skip this and the next 3 steps if you do not use a database server*.
+* Build the project and open the output folder (*bin\Debug*) in the file manager. Locate the application executable file and run two application instances.
+* Double click a record in one window to open the edit dialog, and double click the same record in another window.
+* Type something and save changes in each window. The second time the application should crash and show this error message: *Cannot persist the object. It was modified or deleted (purged) by another application*. The next section demonstrates how to handle such situations gracefully.
```

Comment:
gracefully? Why do you add this? Quantify 'gracefully'.

Reply:
191b6f0645c8f7ee66098afcca0f76d7d88237b5